### PR TITLE
bpo-43574: Dont overallocate list literals

### PR DESCRIPTION
**Pull request by chadnetzer**
Before v3.9, list literals weren't over-allocated, but this behavior regressed starting with v3.9:

Switched to initializing list literals w/ LIST_EXTEND
https://bugs.python.org/issue39320
https://github.com/python/cpython/pull/17984

Commit where over-allocation of list literals first appeared
https://bugs.python.org/issue38328
https://github.com/python/cpython/pull/17114
https://github.com/python/cpython/commit/6dd9b64770af8905bef293c81d541eaaf8d8df52

This changes to the original behavior of not over-allocating lists for list literals, by not over-allocating any 0-length list that is extended.  Technically this is a change in behavior, as lists that are initialized to length 0, and then immediately extended, will not be over-allocated on that initial extension.  This change in behavior is likely innocuous, and possibly desirable, whereas the long-standing Python behavior of not using over-allocation for list literals is almost certainly desired.

Adds a test_overallocation test for lists to find regressions (assuming the regression doesn't alter the behavior of both the list-literal and list-initialization from a known length).

https://bugs.python.org/issue43574

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@chadnetzer

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/test/test_list.py
Comment (pitrou) [inline Lib/test/test_list.py]: Is there a reason you test with two different hardcoded literals?

**Review by pitrou — CHANGES_REQUESTED** [review]
The proposed approach looks fine to me. Two things:
1. Can you sign the contributor's agreement?
2. Can you add a NEWS item? See https://devguide.python.org/committing/?highlight=blurb#updating-news-and-what-s-new-in-python

File: Lib/test/test_list.py
Comment (chadnetzer) [reply Lib/test/test_list.py]: It's really just a minimal set of test conditions to show this specific regression; the behavior change starts with len-3 literals, as len-2 and below are special-cased to use BUILD_LIST: https://github.com/python/cpython/blob/9a50ef43e42ee32450a81ce13ed5a0729d3b84e8/Python/compile.c#L3721

I'll see if I can generalize those tests a bit more without going overboard; probably to test that initializing a new list from a known-length iterable doesn't over-allocate (ie. catching this if it regresses again for other cases), and also maybe confirming over-allocation when expected.

I'll also take care of requests 1 and 2 (CLA and NEWS item)

File: Objects/listobject.c
Comment (sweeneyde) [inline Objects/listobject.c]:
```suggestion
    }
    else {
```
https://www.python.org/dev/peps/pep-0007/#code-lay-out

File: Objects/listobject.c
Comment (sweeneyde) [inline Objects/listobject.c]: Per PEP 7: "All new C code requires braces."

**Comment by chadnetzer** [comment]
I have made the requested changes; please review again.

Let me know if the test_overallocation tests (Lib/test/test_list.py) seem reasonable, or overkill.

**Review by pitrou — APPROVED** [review]
+1, thank you

**Comment by pitrou** [comment]
I'll let other people review before merging this.

File: Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst
Comment (pablogsal) [inline Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst]:
```suggestion
``list` objects don't overallocate when starting empty or are set to be empty. This effectively restores previous ``list`` memory behavior for lists initialized from literals.
```

Comment (chadnetzer) [reply Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst]: > ``list`

Minor typo - missing backtick in first \`\`list\`\`

Comment (chadnetzer) [reply Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst]: Added a commit to adopt this NEWS description (slightly modified)

**Comment by chadnetzer** [comment]
Hmmm.  The `list_resize()` function is used elsewhere in Objects/listobject.c, and in particular is called on lists (empty or otherwise) when inserting, appending, etc.  So this PR will have an effect on empty lists in those areas as well, not just for `list_extend()`.  This will essentially delay overallocation for empty lists in those situations (until after the first insert, append, etc), and possibly affect the amount of overallocation afterwards.

In the interest of fairness, I'll gather some examples to show the behavior change, at least to help decide if any of those situations will suffer a negative consequence as a result. (I'll edit this comment to with the examples, when ready)

**Comment by chadnetzer** [comment]
I made a small but significant tweak to my original proposal, and added a lot of of commentary to the bug tracker issue.  Basically, the latest tweak restores the existing list append/insert behavior for empty lists, while still reverting back to the long-standing behavior of not overallocating list-literals (which, for length 3 or more, are created by using the list_extend() helper)

**Review by sweeneyde — DISMISSED** [review]

File: Lib/test/test_list.py
Comment (pitrou) [inline Lib/test/test_list.py]: This comment says "verified later", but I see no specific test of no-over-allocation of list literals. There is `SizeofTest` in `test_sys.py` but it only tests two literal lengths.
I would imagine this kind of test:
```python
prev_literal = None
for literal in [[] + test_literals]:
    if prev_literal is not None
        # Allocation of literal N uses one more pointer than allocation of literal N-1
        self.assertEqual(sizeof(literal), sizeof(prev_literal) + struct.calcsize('P'))
    prev_literal = literal
```

**Review by pitrou — COMMENTED** [review]
Thanks for the update. I would welcome a bit more thorough testing.

File: Lib/test/test_list.py
Comment (chadnetzer) [reply Lib/test/test_list.py]: Yes, I should have said "verified below", ie. in same test, after the loop.  It's an implicit test, though, since with the current over-allocation strategy, the list capacity can sometimes equal list length, so it's not universally true that each list-literal will have less memory allocated than a normal list of equivalent length.  So it only can test that at least sometimes the list-literals are created with less capacity then a grown list of equivalent length.

I was avoiding a direct test of memory usage, like you propose, just to keep things general.  But I can certainly add such a direct test, which (afaict) is really the only way to test that all the list-literals are being created without over-allocation.

Comment (chadnetzer) [reply Lib/test/test_list.py]: @pitrou I ran with your idea, and added a direct test of the memory allocated for a list-literal, which also let me remove the confusing wording about verifying "later", since that's no longer the case.  Thanks for the suggestion.

File: Lib/test/test_list.py
Comment (gpshead) [inline Lib/test/test_list.py:246]: add `msg=repr(overalloc_amts)` to aid debugging if this fails.

File: Lib/test/test_list.py
Comment (gpshead) [inline Lib/test/test_list.py:230]: add a `msg=repr(literal)` arg to all of the asserts in this loop to aid debugging directly from a test failure log.

File: Lib/test/test_list.py
Comment (gpshead) [inline Lib/test/test_list.py:226]: no action required: meta: this is where I really wish we had an `absltest.paramtereize` or `pytest.mark.parameterize` available in the stdlib.  it'd run all of these as unique tests with appropriate names instead of a loop that prevents others from running upon first failure.

File: Lib/test/test_list.py
Comment (gpshead) [inline Lib/test/test_list.py:248]: Turn this bit into an additional test method.

File: Lib/test/test_list.py
Comment (gpshead) [inline Lib/test/test_list.py:207]: make this set of initial assertions into its own test method.

File: Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst
Comment (gpshead) [inline Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst:2]: quantify what "previous" means.  3.8 and earlier I believe.

**Review by gpshead — CHANGES_REQUESTED** [review]
Overall good, some test maintainability and debugging suggestions.

File: Lib/test/test_list.py
Comment (sweeneyde) [reply Lib/test/test_list.py:226]: I think unittest.subTest() displays parameters on the failing subtests and keeps running through each even on failure.

**Comment by methane** [comment]
> Is this PR still relevant since #31816 was merged?

I think no. This PR and the #87740 can be closed.

**Pull request closed without merging**
